config_proc_macro/src/item_enum.rs

```diff
@@ -0,0 +1,177 @@
+use proc_macro2::TokenStream;
+use quote::quote;
+
+use crate::attrs::*;
+use crate::utils::*;
+
+type Variants = syn::punctuated::Punctuated<syn::Variant, syn::Token![,]>;
+
+pub fn define_config_type_on_enum(em: &syn::ItemEnum) -> syn::Result<TokenStream> {
+    let syn::ItemEnum {
+        vis,
+        enum_token,
+        ident,
+        generics,
+        variants,
+        ..
+    } = em;
+
+    let mod_name_str = format!("__define_config_type_on_enum_{}", ident);
+    let mod_name = syn::Ident::new(&mod_name_str, ident.span());
+    let variants = fold_quote(variants.iter().map(process_variant), |meta| quote!(#meta,));
+
+    let impl_doc_hint = impl_doc_hint(&em.ident, &em.variants);
+    let impl_from_str = impl_from_str(&em.ident, &em.variants);
+    let impl_serde = impl_serde(&em.ident, &em.variants);
+    let impl_deserialize = impl_deserialize(&em.ident, &em.variants);
+
+    Ok(quote! {
+        #[allow(non_snake_case)]
+        mod #mod_name {
+            #[derive(Debug, Copy, Clone, Eq, PartialEq)]
+            pub #enum_token #ident #generics { #variants }
+            #impl_doc_hint
+            #impl_from_str
+            #impl_serde
+            #impl_deserialize
+        }
+        #vis use #mod_name::#ident;
+    })
+}
+
+/// Remove attributes specific to `config_proc_macro` from enum variant fields.
+fn process_variant(variant: &syn::Variant) -> TokenStream {
+    let metas = variant
+        .attrs
+        .iter()
+        .filter(|attr| !is_doc_hint(attr) && !is_config_value(attr));
+    let attrs = fold_quote(metas, |meta| quote!(#meta));
+    let syn::Variant { ident, fields, .. } = variant;
+    quote!(#attrs #ident #fields)
+}
+
+fn impl_doc_hint(ident: &syn::Ident, variants: &Variants) -> TokenStream {
+    let doc_hint = variants
+        .iter()
+        .map(doc_hint_of_variant)
+        .collect::<Vec<_>>()
+        .join("|");
+    let doc_hint = format!("[{}]", doc_hint);
+    quote! {
+        use crate::config::ConfigType;
+        impl ConfigType for #ident {
+            fn doc_hint() -> String {
+                #doc_hint.to_owned()
+            }
+        }
+    }
+}
+
+fn impl_from_str(ident: &syn::Ident, variants: &Variants) -> TokenStream {
+    let vs = variants
+        .iter()
+        .filter(|v| is_unit(v))
+        .map(|v| (config_value_of_variant(v), &v.ident));
+    let if_patterns = fold_quote(vs, |(s, v)| {
+        quote! {
+            if #s.eq_ignore_ascii_case(s) {
+                return Ok(#ident::#v);
+            }
+        }
+    });
+    quote! {
+        impl ::std::str::FromStr for #ident {
+            type Err = &'static str;
+
+            fn from_str(s: &str) -> Result<Self, Self::Err> {
+                #if_patterns
+                return Err("Bad variant");
```

Comment:
Maybe add the argument `s` to the returned error in order to have a nicer message.